### PR TITLE
Prevent R&R line rerenders

### LINKED ISSUE #4568: R&R Forms perfomance
## What went wrong? 😲

When you have a number of rows in an R&R form the performance degrades

## Expected behaviour 🤔
Editing a single line should be pretty snappy

## How to Reproduce 🔨

Steps to reproduce the behaviour:

1. Create an R&R form using a master list with 50+ rows
2. Edit a row on a tablet, or
3. add a `console.info` to the RnRFormLine component

## Your environment 🌱

- Open mSupply Version:
- Legacy mSupply Central Server Version:

- Platform:

- Database type:

### PR DESCRIPTION
**Pull request by mark-prins**
Fixes #4568

# 👩🏻‍💻 What does this PR do?
The `RnRFormLine` component is rendered 3x for every row, on every update.
This is sluggish on android and increasingly so the more lines you have.

I've hobbled the dirty state, as this contributes a render cycle - now the blue dot shows, but there's no global state, so there's no info snack telling you to update. I figure that if you are editing, then the row isn't confirmed and so the finalise button will still prompt.

I had also moved the `useConfirmOnLeaving` hook into the footer as this adds a render cycle when it updates. This is now on every row, and while I'm worried that this adds overhead, it seems to run ok.

Have added the loading indicator to the row. 

With the changes, when you edit a row the row itself renders - previously all rows were re-rendered. 
When you confirm, there are 2 renders for every row, so there's an improvement, even if not ideal.

Here's my old lenovo coping with 50 rows:

https://github.com/user-attachments/assets/91fbc589-6744-457b-ab9e-71991cc3ebf6

## 💌 Any notes for the reviewer?

# 🧪 Testing

- [ ] try editing a medium/large form and get the vibes
- [ ] does it prompt if you navigate away after editing?
- [ ] does it warn you when finalising, if you have un-committed rows?

# 📃 Documentation

- [x] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: 
  1.
  2.

### CODE REVIEW

**Comment by andreievg** [comment]
Just had a quick look at the code, I think it's worth to see what happens if you try to navigate away when multiple lines are dirty, it's looking to me like it would show multiple confirmation modals

**Comment by lache-melvin** [comment]
> does it warn you when finalising, if you have un-committed rows?

It warns that some lines are unconfirmed, but then allows you to continue. I think if we're backing out the dirty state changes, we should at least make that warning text a bit more dramatic - that if you have any unsaved changes, they will be LOST

Also, if i have dirty rows, and then finalise, I still get the confirm on leaving warnings - these should be cleared somehow I think! Or not used...

>  it's looking to me like it would show multiple confirmation modals

Yup, this is happening for me

Can we pause this PR until the tech chat tomorrow? Aware there's already been a bit of effort sunk into this, but need it to be usable for this release... I'm kind of curious to try the zustand slice idea?

**Comment by lache-melvin** [comment]
It's definitely a performance improvement though, thanks for looking into it Mark!

File: server/omsupply-chc-baoknana.sqlite.zip
Comment (lache-melvin) [inline server/omsupply-chc-baoknana.sqlite.zip:1]: maybe remove 😁

File: client/packages/requisitions/src/RnRForms/DetailView/ContentArea.tsx [hunk 1]
```diff
@@ -47,36 +48,55 @@ export const ContentArea = ({
...
             borderRight: '1px solid blue',
+            left: 72,
+            minWidth: '300px',
+            position: '-webkit-sticky',
           },
         }}
       />
       <Table
         sx={{
+          height: '500px',
```
Comment (lache-melvin) [inline client/packages/requisitions/src/RnRForms/DetailView/ContentArea.tsx:98]: turns out adding `borderCollapse: 'separate'` here solves the borders not staying sticky on side scroll!

File: client/packages/requisitions/src/RnRForms/DetailView/ContentArea.tsx [hunk 2]
```diff
@@ -47,36 +48,55 @@ export const ContentArea = ({
...
             width: 80,
           },
           '.second-column': {
-            position: '-webkit-sticky',
-            left: 88,
-            minWidth: '300px',
             borderRight: '1px solid blue',
+            left: 72,
+            minWidth: '300px',
+            position: '-webkit-sticky',
```
Comment (lache-melvin) [inline client/packages/requisitions/src/RnRForms/DetailView/ContentArea.tsx:92]: I think we might need the column widths...

https://github.com/user-attachments/assets/0e042b4c-c2a9-48de-b981-38f0173b6799

File: client/packages/requisitions/src/RnRForms/DetailView/ContentArea.tsx (hunk 1 above)
Comment (mark-prins) [reply client/packages/requisitions/src/RnRForms/DetailView/ContentArea.tsx:98]: ❤️  legend!!

**Review by lache-melvin — APPROVED** [review]
Looks great Mark! Working well for me, just the column widths that I'm concerned about.

And obviously the changes clearing when you scroll, but all good with that in a separate PR 🙏

**Pull request merged**